Add Best Practice: Naming Root Projects (#34468)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_general.adoc

=== Tags

<<tags_reference.adoc#tag:properties,`#properties`>>

[[name_your_root_project]]
== Name Your Root Project

Always name your root project in the `settings.gradle(.kts)` file.

=== Explanation

> While an empty `settings.gradle(.kts)` file is all that is minimally necessary for a multi-project build, you should always set the `rootProject.name` property in it.

Reviewer: ```suggestion
While an empty `settings.gradle(.kts)` file is enough to create a multi-project build, you should always set the `rootProject.name` property.
```

> Without setting a name explicitly, the root project's name implicitly defaults to the name of the directory containing the build.

Reviewer: ```suggestion
By default, the root project’s name is taken from the directory containing the build.
```

> Some directory names may not be valid or suitable as project names, such as those containing spaces, Gradle logical path separators or other special characters.

Reviewer: ```suggestion
This can be problematic if the directory name contains spaces, Gradle logical path separators, or other special characters.
```

> The paths to tasks in build should be able to be specified reliably regardless of where the build lives.

Reviewer: ```suggestion
It also makes task paths dependent on the directory name, rather than being reliably defined.
```

> Explicitly setting the root project's name also ensures that it referred to consistently.

Reviewer: ```suggestion
Explicitly setting the root project's name ensures consistency across environments.
```

> Project names are used in various places, such as error messages, logs, and reports.

Reviewer: ```suggestion
Project names appear in error messages, logs, and reports, and builds often run on different machines, such as CI servers.
```

> Builds may execute on a variety of machines or environments, such as CI servers, and should report the same root project name anywhere to make the project more comprehensible.

Reviewer: ```suggestion
Defining the name in `settings.gradle(.kts)` guarantees the same name everywhere, making the build easier to understand and maintain.
```

=== Example

==== Don't Do This

> ====

Reviewer 2: This is kind of a useless example that is covered by `In this build, the settings file is empty and the root project is unnamed.`

I say we drop the example as it doesn't help anyone.

Reviewer: I think we can keep this "Don't Do This" example IF we had an example of the `projects` report below.

include::sample[dir="snippets/bestPractices/nameYourRootProject-avoid/kotlin",files="settings.gradle.kts[]"]
include::sample[dir="snippets/bestPractices/nameYourRootProject-avoid/groovy",files="settings.gradle[]"]
====

> In this build, the settings file is empty and the root project is unnamed.

Reviewer: ```suggestion
In this build, the settings file is empty and the root project has no explicit name.
```

> If you try running the `projects` report, you will see an implicit name used for the root project that is based on the build's current location.

Reviewer: ```suggestion
Running the `projects` report shows that Gradle assigns an implicit name to the root project, derived from the build’s current directory.
```

Reviewer: Would be nice to see an output here of `projects` -> `nameYourRootProject-avoid.out`
